### DISCUSSION TITLE
DALY BMS shows inverted current

### DISCUSSION THREAD

**Discussion by HM-42** [Louisvdw/dbus-serialbattery#461] [Q&A] [answered]
Hi,
i am using a DALY 150A BMS / 16S with a Cerbo and Multiplus-II 3000.
The driver works fine so far, but in the remote console, the current shows negative if charged and the BMS 
says "discharging"
<img width="505" alt="image" src="https://user-images.githubusercontent.com/96735945/218751317-68f5eb42-46ca-4116-8da7-d1251b4b6ba1.png">

<img width="395" alt="image" src="https://user-images.githubusercontent.com/96735945/218751797-57161199-ea80-4de1-bdc4-73d9c5e2abbc.png">

<img width="651" alt="image" src="https://user-images.githubusercontent.com/96735945/218752458-d08fa099-42b1-481b-9b8a-7729fd05fff6.png">

any ideas what to change ?

**Comment by Louisvdw** [comment, marked as answer]
Hi
Some Daly BMS has the value swopped around. You can fix it in the settings in utils.py 
Look for INVERT_CURRENT_MEASUREMENT

**Reply by HM-42** [reply]
Thanks for the quick reply!
I will try tomorrow....

**Reply by HM-42** [reply]
I found it, but "utils.py" is not the file to change.
should be changed in "default_config.ini"


<img width="579" alt="image" src="https://user-images.githubusercontent.com/96735945/218772045-f2ad0188-d032-4623-99d0-0aeb0ed98107.png">

**Reply by mrgithub** [reply]
I have the same problem and a brand new user of Victron and this script.

Where is default_config.ini file as I don't see it in /data/etc/dbus-serialbattery/ ?

Instead I found it in utils.py as Louis described and changed the value to -1.   

Rebooted and the VRM shows the energy flowing in correct direction and no negative values.  The voltage is 53.28 and system is charging around ~4A (~200w).  Why is it so low?

**Reply by mr-manuel** [reply]
@HM-42  `utils.py` was the file to go until including `0.14.3`. You probably downloaded a copy of the repository and not the latest version?

In future versions you have to create a `config.ini` with your custom values, the `default_config.ini` else will be overridden on the next update.

**Reply by mrgithub** [reply]
@mr-manuel I don't see this config.ini in  /data/etc/dbus-serialbattery/ and I have 0.14.3

Also battery charging so slooow - why? 

<img width="683" alt="image" src="https://user-images.githubusercontent.com/3989211/219029481-9456a0f2-bb65-4f1c-8eef-0dcad0fa0137.png">

The AC loads are parallel to the inverter and using external current sensor, I don't know why VRM shows inverter in the middle of it.

**Reply by mr-manuel** [reply]
@mrgithub that's correct. The `0.14.3` is the last release without it.

What do you see in Venus OS under SerialBattery -> Parameters?

**Reply by HM-42** [reply]
i just tried it own system (was on a holiday trip the last two days):
I changed in utils.py and it worked.
I also changed the BATTERY_CAPACITY to my values (230Ah) and it also works fine.

**Comment by mrgithub** [comment]
So 0.14.3 relies solely on utils.py and config.ini is deprecated - OK

Hardware setup:
- Multiplus 48v 5000 (70A charging capability)
- Daly 250A BMS
- 16 x Eve 304Ah LifeP04 batteries (15kWH of storage)
- External current sensor
- ESS for use in UK
- Tomzen 250A 2 pole DC breaker (35-50kA Icu Ics)

---> So my cables and hardware should easily support 240A charge and discharge should I daisy chain more inverters over time along with solar charge controllers.

Here's the changed lines in utils.py file for my setup (so it could be wrong as I've tried to make it match my hardware's max capacity/bandwidth).


'# Set Steps to reduce battery current. The current will be changed linear between those steps'
'CELL_VOLTAGES_WHILE_CHARGING         = [3.55, 3.50, 3.45, 3.30]'
'MAX_CHARGE_CURRENT_CV                = [0, 2, 100, 240]  #[   0,    2,  30,  60]'

'CELL_VOLTAGES_WHILE_DISCHARGING      = [2.70, 2.80, 2.90, 3.10]'
'MAX_DISCHARGE_CURRENT_CV             = [0, 5, 100, 240]  #[   0,    5,  30,  60]'

'# Set Steps to reduce battery current. The current will be changed linear between those steps'
'TEMPERATURE_LIMITS_WHILE_CHARGING    = [55, 40,  35,   5,  2, 0]'
'MAX_CHARGE_CURRENT_T                 = [0,100,240,240,100,0]   #[ 0, 28, 60, 60, 28, 0]'

'TEMPERATURE_LIMITS_WHILE_DISCHARGING = [55, 40,  35,   5,  0, -20]'
'MAX_DISCHARGE_CURRENT_T              = [0,100,240,240,100,0]   #[ 0, 28, 60, 60, 28,   0]'

'# battery Current limits
'MAX_BATTERY_CHARGE_CURRENT = 240.0     #50.0'
'MAX_BATTERY_DISCHARGE_CURRENT = 240.0  #60.0'

'#charge current limits'
'CC_CURRENT_LIMIT1 = 240'

'#discharge current limits'
'DC_CURRENT_LIMIT1 = 240'

'# Daly settings'
'# Battery capacity (amps) if the BMS does not support reading it'
'BATTERY_CAPACITY = 304'
'# Invert Battery Current. Default non-inverted. Set to -1 to invert'
'INVERT_CURRENT_MEASUREMENT = -1'

The inverter started charging at a higher rate fluctuating between 60-100A according to VRM.  Then it stopped and there's a mismatch between what VRM is reporting and what the serial battery driver.

<img width="493" alt="image" src="https://user-images.githubusercontent.com/3989211/219096298-be305722-7eb4-44b1-8f4b-e3144b4a3229.png">

<img width="688" alt="image" src="https://user-images.githubusercontent.com/3989211/219096835-01c17a48-d928-4bdc-b5c4-9e5d837fbe66.png">

Questions:

1. Is it necessary to change all these values in utils.py or have I over complicated it?
2. Are these settings appropriate for this setup? (temp steps, volt steps, capacity, charge limits, battery limits, etc) 
4. Why are there two different voltages between Multiplus (DC voltage) and Serial battery driver voltage (Cell Sum)?
5. How do I keep it charging until Cells Sum has reached the DC voltage so that it can balance the cells (at 3.45v)?
6. (Side question: why doesn't VRM show system as a grid parallel setup as its using an external current sensor?)

<img width="499" alt="image" src="https://user-images.githubusercontent.com/3989211/219098831-1d1a4868-c70d-4f65-aa62-a3be4aee2581.png">

**Reply by mr-manuel** [reply]
`config.ini` is NOT deprecated, it is just not released yet. It will be introduced with the next version.

**Reply by mrgithub** [reply]
Thanks - do you or @Louisvdw know about the other questions above?

**Reply by mr-manuel** [reply]
1. I would have changed only the highest values. 50 and 60 -> 240, but this is a personal opinion.
2. They should be good. What are your charger setting in the VE Config for the MultiPlus?
3. One is measured by the MultiPlus and one by the BMS. I have no Daly BMS, so I cannot tell how accurate it is. I trust more in the voltage of the MultiPlus.
4. You have to check the settings in the VE Config, additionally you have set the CVL to 55.20 V. Whichever value is lower will be respected. For the driver you can change that here https://github.com/Louisvdw/dbus-serialbattery/blob/3685d15a5d1ba801867bbdb98f57fd805ea2c24a/etc/dbus-serialbattery/utils.py#L89 but I woudn‘t, as increasing the maximum voltage decreases battery life. I changed the starting voltage for balancing to 3.40 in BMS settings.
5. What do you have set unter System -> ESS -> Grid metering?

**Reply by mrgithub** [reply]
1. The Daly BMS 250A 16s model does 250A discharge and 125A charge.  The 304Ah batteries spec sheet says they are rated at 250A so I'm tinkering with this.  Thanks for your opinion.
2. VE Config settings - Don't these get overruled by utils.py?
![image](https://user-images.githubusercontent.com/3989211/219379915-8342320b-d6d6-4b64-bef6-f6e214918598.png)
![image](https://user-images.githubusercontent.com/3989211/219379942-30d17df9-66b3-48cc-8d16-e989bff00ce5.png)
4. I've observed CVL changes over time eg batteries get charge and warm and CVL reduces.  You're suggesting the CVL will be the lowest between VE Config float/absorption voltage then somehow utils.py.  Ok I will need to think about how this dynamic plays out and readjust the VE Config settings.  I'm not sure what they should be set to.  In your example of MAX_CELL_VOLTAGE = 3.45, what do you have your VE Config settings at?
5. If I set it to 
<img width="495" alt="image" src="https://user-images.githubusercontent.com/3989211/219381833-07359a78-d6a6-4a92-836c-bee8b5efeb00.png">
Then VRM looks almost right but AC Loads has now become redundant and everything is a Critical Load. Physically, the inverter only has one connection to the grid via AC In 1 with no other AC connection.
<img width="661" alt="image" src="https://user-images.githubusercontent.com/3989211/219382218-31f6eda3-b66a-4aaf-994a-7cbf533f52e0.png">
Anyway ESS doesn't seem to operate unless 'grid metering' is set to 'charger/inverter' (albeit with external current sensor). I guess its fine as it seems to be only an aesthetic issue.

Thanks so much for answering. I'm learning a lot here from this interaction.

**Comment by mr-manuel** [comment]
1. In this case I would set it to
```
# Set Steps to reduce battery current. The current will be changed linear between those steps
CELL_VOLTAGES_WHILE_CHARGING         = [3.55, 3.50, 3.45, 3.30]
MAX_CHARGE_CURRENT_CV                = [   0,    2,   30,  125]

CELL_VOLTAGES_WHILE_DISCHARGING      = [2.70, 2.80, 2.90, 3.10]
MAX_DISCHARGE_CURRENT_CV             = [   0,    5,   30,  250]

# Set Steps to reduce battery current. The current will be changed linear between those steps
TEMPERATURE_LIMITS_WHILE_CHARGING    = [55, 40,  35,   5,  2, 0]
MAX_CHARGE_CURRENT_T                 = [ 0, 28, 125, 125, 28, 0]

TEMPERATURE_LIMITS_WHILE_DISCHARGING = [55, 40,  35,   5,  0, -20]
MAX_DISCHARGE_CURRENT_T              = [ 0, 28, 250, 250, 28,   0]

# battery Current limits
MAX_BATTERY_CHARGE_CURRENT = 125.0
MAX_BATTERY_DISCHARGE_CURRENT = 250.0
```

2. No, as my understanding it gets the lower value. But your charging settings in VE Config looks nearly like mine.
3. ![grafik](https://user-images.githubusercontent.com/12592765/219398446-805b8b56-255b-4e41-a069-bba77177ecc2.png)
4. Unfortunately I don't know an answer to that

**Reply by mrgithub** [reply]
1. Thanks for sharing your settings and opinions.  I've adjusted the settings to 240A for charge and discharge as the batteries are rated for 250A continuous.  (I have a Tomzen battery DC breaker rated at 250A just after the Daly BMS which itself is rated at 250A discharge and 125A charge).
2. So every week it should go into absorption for 2 hours to balance itself - clever.
3. Why is float value lower than absorption I thought it was supposed to be higher?  (Although I have seen YouTubers say 3 stages of charging is meaningless for LifeP04 batteries).
4. Np - My ESS now seems to be functioning great and charging during the scheduled hours.  Problem now is inverter and SerialBattery driver provide two different system voltages that are about 1v apart but this only occurs when charging after.  During discharge both reported voltages are very similar.  The min max battery voltages sort of reflect the issue by reporting a great disparity while charging then the deviation disappears when ESS starts using the battery after 6am. The readings from each cell look solid (crimped and bolted tight) but perhaps there's a tiny looseness that is emphasised and accumulates into a 1v difference with the warmth of charging - Have you seen graphs like this before?
<img width="814" alt="image" src="https://user-images.githubusercontent.com/3989211/219641322-f0dc516a-f82a-4177-91a2-007aec40335c.png">
 
Your input has greatly excelled my knowledge - thanks again!

**Reply by mr-manuel** [reply]
2. Yes, this is the time my BMS need to correct the difference between all cells. So how long you go into absorption depens on the cell voltage difference (therefore also capacity difference) and the balancing current.
3. Many people, like me, use absorption just for balancing purposes. The float voltage is to not keep the batteries under pressure when they are full. This should expand their lifetime a bit. Float should never by higher than float, that don't make any sense to me.
![grafik](https://user-images.githubusercontent.com/12592765/219652883-e0e8a6ff-497d-4507-99a8-5ff0c0b3344f.png)

**Comment by Marccatamaran** [comment]
Why the 16 cell voltage are in red? 
I have the same issue?

**Reply by mr-manuel** [reply]
This should indicate that balancing is active and working right now.

**Comment by Marccatamaran** [comment]
But they are always red even when I am not charging? Balancing occur only when charging?

**Comment by mrgithub** [comment]
@mr-manuel thanks for responding
@Marccatamaran  I have the same question

I really love this driver so made a small donation.

Here are some more questions for you guys and @Louisvdw :
1. When will it perform the absorption every 7 days using the VE Config settings as above?
2. How to set the Daly balancing start voltage from util.py?
3. It seems, the temp on VRM dashboard doesn't update unless webpage is refreshed
4. Is there a way to log the individual battery cell voltages over time into a csv file to analyse the performance of individual cells? Or even better, log it in a VRM graph?
5. The cells should turn red when they are balancing but it seems all the cells are red all the time in the console. Can the min and max cells be a different colour and can only the cells being balanced be highlighted?

**Comment by greguser** [comment]
Hi everyone

got it thanks to your help, everything works fine:

Multiplus 48v 3000 (35A charging capability)
Daly 250A BMS
16 x 280Ah LifeP04 batteries 48V
Shelly 3EM measurement in CebroGX
ESS on CerboGX
Everything integrated on Homassisant (via Modbus)

I just have one more question.
Since I integrated the Daly Bms on the Cerbo GX, the Multiplus also charges over 35A up to the value of 50A set in utils.py.
At least that's how it is displayed.
According to the Multiplus data sheet, the charging capacity is 35 A. Can there be problems when switching?

Should I rather set the value in utils.py to 35 A?

**Reply by mr-manuel** [reply]
The MultiPlus will automatically stop on it's hardware limit. So it's no problem, if the limit in `utils.py` is higher.

**Reply by greguser** [reply]
Thanks

**Comment by dd6uw** [comment]
Hi, i changed the Value to -1, but it wont work.
where is my mistake?
# Invert Battery Current. Default non-inverted. Set to -1 to invert
INVERT_CURRENT_MEASUREMENT = -1
#int(config["DEFAULT"]["INVERT_CURRENT_MEASUREMENT"]) )

**Reply by dd6uw** [reply]
Now i generated a further problem, cerbo is nor reporting a lost connection to BMS.🤔

**Reply by Louisvdw** [reply]
Hi @dd6uw 
How did you make the change to the config? Look at the troubleshoot section in the documentation to see how to view the log files. Chances are the edit had an error in that is blocking the driver from starting. If you cannot revert, then remove your config and restart should put it back to default.

**Reply by dd6uw** [reply]
Hi Louis,  added -1 in the row with invert_current_Measuerment and commented the other stuff out with #int(config["DEFAULT"]["INVERT_CURRENT_MEASUREMENT"]) )
After the disconnection happend, i rebuild the old  version and the BMS is still disconnected

**Reply by dd6uw** [reply]
i fixed my connection problem by reinstalling.

**Reply by Louisvdw** [reply]
for the latest released version you need to make this change in the /data/etc/dbus-serialbattery/config.ini file.
Use the Nano editor described in https://louisvdw.github.io/dbus-serialbattery/general/install/#settings-locationpath to minimise editing issues. A reinstall would not override this file, so I presume your changes was elsewhere.

**Reply by dd6uw** [reply]
yip , i did it, now, the Charging Current is right.
Before that, i  fought against a new problem, daly sends a hih Voltage Warning with 11.7 Volts in my 12 V system.

**Reply by mr-manuel** [reply]
See https://github.com/Louisvdw/dbus-serialbattery/issues/653
